drivers/sensor/st/iis3dwb/iis3dwb.c

```diff
@@ -0,0 +1,453 @@
+
+	if (err) {
+		rtio_iodev_sqe_err(iodev_sqe, err);
+	} else {
+		rtio_iodev_sqe_ok(iodev_sqe, 0);
+	}
+}
+
+/*
+ * Create a chain of SQEs representing a bus transaction to read a reg.
+ * The RTIO-enabled bus driver will:
+ *
+ *     - write "reg" address
+ *     - read "len" data bytes into "buf".
+ *     - call complete_op callback
+ */
+void iis3dwb_rtio_rd_transaction(const struct device *dev,
+				 uint8_t *regs, uint8_t regs_num,
+				 struct spi_buf *buf,
+				 struct rtio_iodev_sqe *iodev_sqe,
+				 rtio_callback_t complete_op_cb)
+{
+	struct iis3dwb_data *iis3dwb = dev->data;
+	struct rtio *rtio = iis3dwb->rtio_ctx;
+	struct rtio_iodev *iodev = iis3dwb->iodev;
+	struct rtio_sqe *write_addr;
+	struct rtio_sqe *read_reg;
+	struct rtio_sqe *complete_op;
+	uint8_t i;
+
+	for (i = 0; i < regs_num; i++) {
+
+		write_addr = rtio_sqe_acquire(rtio);
+		read_reg = rtio_sqe_acquire(rtio);
+
+		if (write_addr == NULL || read_reg == NULL) {
+			return;
+		}
+
+		regs[i] |= 0x80; /* mark the SPI read transaction */
+
+		rtio_sqe_prep_tiny_write(write_addr, iodev, RTIO_PRIO_NORM, &regs[i], 1, NULL);
+		write_addr->flags = RTIO_SQE_TRANSACTION;
+		rtio_sqe_prep_read(read_reg, iodev, RTIO_PRIO_NORM, buf[i].buf, buf[i].len, NULL);
+		read_reg->flags = RTIO_SQE_CHAINED;
+	}
+
+	complete_op = rtio_sqe_acquire(rtio);
+	if (complete_op == NULL) {
+		return;
+	}
+
+	rtio_sqe_prep_callback_no_cqe(complete_op, complete_op_cb, (void *)dev, iodev_sqe);
+
+	rtio_submit(rtio, 0);
+}
+
+static void iis3dwb_submit_one_shot(const struct device *dev, struct rtio_iodev_sqe *iodev_sqe)
+{
+	const struct sensor_read_config *cfg = iodev_sqe->sqe.iodev->data;
+	const struct sensor_chan_spec *const channels = cfg->channels;
+	const size_t num_channels = cfg->count;
+	uint32_t min_buf_len = sizeof(struct iis3dwb_rtio_data);
+	uint64_t cycles;
+	int rc = 0;
+	uint8_t *buf;
+	uint32_t buf_len;
+	struct iis3dwb_rtio_data *edata;
+	struct iis3dwb_data *data = dev->data;
+
+	/* Get the buffer for the frame, it may be allocated dynamically by the rtio context */
+	rc = rtio_sqe_rx_buf(iodev_sqe, min_buf_len, min_buf_len, &buf, &buf_len);
+	if (rc != 0) {
+		LOG_ERR("Failed to get a read buffer of size %u bytes", min_buf_len);
+		return;
+	}
+
+	edata = (struct iis3dwb_rtio_data *)buf;
+
+	edata->has_accel = 0;
+	edata->has_temp = 0;
+
+	rc = sensor_clock_get_cycles(&cycles);
+	if (rc != 0) {
+		LOG_ERR("Failed to get sensor clock cycles");
+		rtio_iodev_sqe_err(iodev_sqe, rc);
+		return;
+	}
+
+	edata->header.is_fifo = false;
+	edata->header.range = data->range;
+	edata->header.timestamp = sensor_clock_cycles_to_ns(cycles);
+
+	for (int i = 0; i < num_channels; i++) {
+		switch (channels[i].chan_type) {
+		case SENSOR_CHAN_ACCEL_X:
+		case SENSOR_CHAN_ACCEL_Y:
+		case SENSOR_CHAN_ACCEL_Z:
+		case SENSOR_CHAN_ACCEL_XYZ:
+			edata->has_accel = 1;
+
+			uint8_t outx_regs[]  = { IIS3DWB_OUTX_L_A, };
+			struct spi_buf buf_xl[] = { {(uint8_t *)edata->accel, 6} };
```

Comment:
We shouldn't be using the memory from the driver here, the user should be providing the data with the SQE. If we have the MPU on, this would cause a fault.